compiler/rustc_hir_analysis/src/collect.rs

```diff
@@ -1232,17 +1239,10 @@ fn infer_return_ty_for_fn_sig<'tcx>(
                     "for more information on `Fn` traits and closure types, see \
                      https://doc.rust-lang.org/book/ch13-01-closures.html",
                 );
-            } else if let Some(i_ty) = suggest_impl_iterator(tcx, ret_ty, ty.span, hir_id, def_id) {
-                diag.span_suggestion(
-                    ty.span,
-                    "replace with an appropriate return type",
-                    format!("impl Iterator<Item = {}>", i_ty),
-                    Applicability::MachineApplicable,
-                );
             }
             diag.emit();
 
-            fn_sig
+            ty::Binder::dummy(fn_sig)
```

Comment:
And I think this might be dead code now, right?

Reply:
Wait, why's it dead code?

Reply:
Because it only triggers if `ret_ty.is_closure()`, but if it is then the structured suggestion would be given. I think that the only way this would be triggered is if you had something like

```
fn foo() -> _ {
    || foo()
}
```
where the resolved type ends up being something like `impl Fn() -> _` or `impl Fn() -> [type error]`.

Reply:
Oh, you mean the `if ..is_closure()` branch? That'll still trigger when we have a non-suggestable return type too, so I'd prefer to keep it.